Add downgrade test to CI

aanm opened the pull request:
Rewrite upgrade test to do a clean downgrade; then an upgrade with a previous state; after that do a downgrade with the previous state as well. When the tests are completed, do perform a state clean to avoid the the upgrade test to affect remaining tests.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5330)
<!-- Reviewable:end -->

aanm committed:
test/helpers: add CiliumInstallVersion to install specific cilium version

This allows to deploy a specific Cilium version from k8s descriptors
available on github.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
test: test an upgrade and a downgrade of a particular Cilium version

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
test: test against branches and not tags

As we didn't release any 1.0, 1.1 and 1.2 after the necessary k8s
descriptor fixes we need to perform the upgrade and downgrade tests
against those branches instead of a specific Cilium version.

Signed-off-by: André Martins <andre@cilium.io>

eloycoto commented on test/k8sT/Updates.go:
```diff
@@ -125,6 +155,34 @@ func ValidateCiliumUpgrades(kubectl *helpers.Kubectl) (func(), func()) {
 
+		validateEndpointsConnection := func() {
+			err := kubectl.CiliumEndpointWaitReady()
+			Expect(err).To(BeNil(), "Endpoints are not ready after timeout")
```
ExpectWithOffset

aanm replied:
Done

eloycoto commented on test/k8sT/Nightly.go:
```diff
@@ -420,7 +424,15 @@ var _ = Describe("NightlyExamples", func() {
+				"cilium-cm-patch-clean-cilium-state.yaml",
+				version,
+			)
+			Expect(err).To(BeNil(), fmt.Sprintf("Cilium %s was not able to be deployed", version))
```
That `fmt.Sprintf` is not needed

aanm replied:
Fixed

eloycoto commented on test/k8sT/Nightly.go:
```diff
@@ -433,12 +445,11 @@ var _ = Describe("NightlyExamples", func() {
 			func(image string) {
 				It(fmt.Sprintf("Update Cilium from %s to master", image), func() {
-					helpers.InstallExampleCilium(kubectl, image)
+					By("Cleaning up everything before testing an upgrade")
```
I would add the image in that message, due there is an image, and not sure what image it is.

aanm replied:
Done

eloycoto commented on test/k8sT/Nightly.go:
```diff
@@ -433,12 +445,11 @@ var _ = Describe("NightlyExamples", func() {
-					Expect(err).To(BeNil(), "Endpoints are not ready after timeout")
 					var assertUpgradeSuccessful func()
-					assertUpgradeSuccessful, cleanupCallback = ValidateCiliumUpgrades(kubectl)
+					assertUpgradeSuccessful, cleanupCallback = ValidateCiliumUpgrades(kubectl, image, "k8s1:5000/cilium/cilium-dev:latest")
```
`k8s1:5000/cilium/cilium-dev:latest` should be a const in helpers.

aanm replied:
Done

eloycoto commented on test/k8sT/Updates.go:
```diff
@@ -11,6 +11,20 @@ import (
 )
 
 var _ = Describe("K8sUpdates", func() {
+
```
You can do whatever you want, but my advice here wil be the following: 

- Have two separate test, if not.. the test that always fail is the Upgrade one, and It'll be a mess. 
- That description in the top of the describe can be done with By messages, and people will understand the test as it runs. 

But as I said, that can be kept, and feel free to merge this one. 
Regards.

aanm replied:
The I've added more By's. I'll keep this description as it is easier to have an overview of what the test is suppose to do instead of reading all 250 lines.

WRT the 2 tests, it's really difficult to split both tests as one depends on the other. The cilium state present after the upgrade should and needs to affect the downgrade test.

eloycoto replied:
I mean, as I said, I'm happy to keep as you think that it's better, but I also read few weeks back(In slack)you said that the downgrade test can fail, but the update will never fail, so we don't really now if the update works correct and we can have a lot of false positives. 

TBH: I would keep the Update test as it is, and add a extensive and longer test (With more policies, CIDR policies, background connectivity, different pods with init containers) in the nightly test, that will help us to be close to the users world, with more nodes ( Nightly has 4 nodes) , different services, etc.. 

Regards

eloycoto commented on test/k8sT/Updates.go:
```diff
@@ -70,31 +84,38 @@ var _ = Describe("K8sUpdates", func() {
 	})
 
 	BeforeEach(func() {
-		kubectl.Exec("sudo docker rmi cilium/cilium")
```
There is a reason for this, if not the `cilium/cilium` image is  when we do `make docker-image`, now with docker.io prefix+tags will be ok, but maybe I would keep it.

aanm replied:
Discussed offline

eloycoto commented on test/helpers/kubectl.go:
```diff
@@ -829,16 +831,50 @@ func (kub *Kubectl) CiliumInstall(dsPatchName, cmPatchName string) error {
+		return filepath.Join(BasePath, "k8sT", "manifests", filename)
+	}
+	getK8sDescriptor := func(filename string) string {
+		return fmt.Sprintf("https://raw.githubusercontent.com/cilium/cilium/%s/examples/kubernetes/%s/%s", versionTag, GetCurrentK8SEnv(), filename)
```
I'm not sure, but I can't get why this is needed at all, could you explain?

aanm replied:
@eloycoto It is the official way we have access to the kubernetes descriptors for the `versionTag`.

eloycoto commented on test/k8sT/Updates.go:
```diff
@@ -125,6 +155,34 @@ func ValidateCiliumUpgrades(kubectl *helpers.Kubectl) (func(), func()) {
+			ExpectKubeDNSReady(kubectl)
+
+			err = kubectl.WaitForKubeDNSEntry(app1Service, helpers.DefaultNamespace)
+			Expect(err).To(BeNil(), "DNS entry is not ready after timeout")
```
ExpectWithOffset

aanm replied:
Done

eloycoto commented on test/k8sT/Updates.go:
```diff
@@ -125,6 +155,34 @@ func ValidateCiliumUpgrades(kubectl *helpers.Kubectl) (func(), func()) {
+			Expect(err).To(BeNil(), "DNS entry is not ready after timeout")
+
+			err = kubectl.CiliumEndpointWaitReady()
+			Expect(err).To(BeNil(), "Endpoints are not ready after timeout")
```
ExpectWithOffset

aanm replied:
Done

eloycoto commented on test/k8sT/Updates.go:
```diff
@@ -125,6 +155,34 @@ func ValidateCiliumUpgrades(kubectl *helpers.Kubectl) (func(), func()) {
+			appPods := helpers.GetAppPods(apps, helpers.DefaultNamespace, kubectl, "id")
+
+			err = kubectl.WaitForKubeDNSEntry(app1Service, helpers.DefaultNamespace)
+			Expect(err).To(BeNil(), "DNS entry is not ready after timeout")
```
ExpectWithOffset

aanm replied:
Done

eloycoto commented on test/k8sT/Updates.go:
```diff
@@ -125,6 +155,34 @@ func ValidateCiliumUpgrades(kubectl *helpers.Kubectl) (func(), func()) {
+			res := kubectl.ExecPodCmd(
+				helpers.DefaultNamespace, appPods[helpers.App2],
+				helpers.CurlFail("http://%s/public", app1Service))
+			res.ExpectSuccess("Cannot curl app1-service")
```
ExpectWithOffset

aanm replied:
Done

eloycoto commented on test/k8sT/Updates.go:
```diff
@@ -125,6 +155,34 @@ func ValidateCiliumUpgrades(kubectl *helpers.Kubectl) (func(), func()) {
+			res = kubectl.ExecPodCmd(
+				helpers.DefaultNamespace, appPods[helpers.App2],
+				helpers.CurlFail("http://%s/private", app1Service))
+			res.ExpectFail("Expect a 403 from app1-service")
```
ExpectWithOffset

aanm replied:
Done

aanm committed:
test: check for pods marked to be deleted in helpers

The WaitforPods is a helper with a timeout so it should bail out in the
when checking for pre-conditions.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
test: remove cilium/cilium:latest after building it

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
test: add specific k8s manifests for each Cilium version

As Cilium versions differ in the arguments accepted between versions we
need to have a specific DaemonSet patch for each Kubernetes version
supported.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
test: prevent upgrade / downgrade test in unsupported k8s versions

As older versions of Cilium can stop supporting new k8s versions, the
tests should only run if Cilium supports running in a specific k8s
version. If any of the Cilium versions to be tested are unsupported,
the upgrade / downgrade test is skipped.

Signed-off-by: André Martins <andre@cilium.io>

ianvernon commented on test/helpers/cons.go:
```diff
@@ -171,10 +171,9 @@ const (
-
-	configMap = "ConfigMap"
-	daemonSet = "DaemonSet"
+	CiliumStableVersion      = "v1.1.4"
```
This should be updated to v1.2.1

aanm replied:
Agreed, but that should be done in a different PR.

ianvernon commented on test/k8sT/Updates.go:
```diff
@@ -11,6 +11,20 @@ import (
+	// This test runs 8 steps as following:
+	// 1 - delete all pods. Clean cilium, this can be, and should be achieved by
+	// `clean-cilium-state: "true"` option that we have in configmap
+	// 2 - install cilium `cilium:v1.1.4`
```
Nit: this comment will go out of date quickly once we update the existing latest released version. What about "the latest version specified in `test/helpers/cons.go:CiliumStableVersion`"?

aanm replied:
This should be an overview of what the test is doing, it's easier to understand the test logic that read the test line by line.

ianvernon commented on test/k8sT/Updates.go:
```diff
@@ -11,6 +11,20 @@ import (
+	// 5 - make endpoints talk with each other with policy
+	// 6 - downgrade cilium to `cilium:v1.1.4`
+	// 7 - make endpoints talk with each other with policy
+	// 8 - delete all pods. Clean cilium, this can be, and should be achieved by
```
By reading this, it makes it sound like as if we don't upgrade back to `k8s1:5000/cilium/cilium-dev:latest`?

aanm replied:
You're correct. I'm assuming all other tests will make sure if Cilium is up and running. Which the BeforeAll does that already. Should we upgrade back to `k8s1:5000`? Maybe we should. I'll add to the afterall

aanm replied:
Done

ianvernon replied:
Thanks!

ianvernon commented on test/k8sT/Updates.go:
```diff
@@ -61,31 +75,38 @@ var _ = Describe("K8sUpdates", func() {
 	})
 
-	It("Updating Cilium stable to master", func() {
+	It("Updating and Downgrading Cilium stable to master", func() {
```
Nit: change to "Tests upgrade and downgrade"?

aanm replied:
Done

ianvernon commented on test/helpers/cons.go:
```diff
@@ -171,7 +171,7 @@ const (
 	KubectlPolicyNameSpaceLabel = k8sConst.PolicyLabelNamespace
 
-	CiliumStableVersion      = "v1.1.4"
+	CiliumStableVersion      = "v1.1"
```
Nice, this makes things easier!

aanm replied:
Don't get too excited, check the commit message.

ianvernon commented on test/k8sT/manifests/1.10/v1.0/cilium-ds-patch.yaml:
```diff
@@ -0,0 +1,15 @@
+---
```
I'm wary of this... is there a reason why we can't use patches from the v1.0, v1.1, v1.2 repos themselves? Also, how will these be updated if the generated YAMLs are updated as well with new arguments, etc.?

aanm replied:
> is there a reason why we can't use patches from the v1.0, v1.1, v1.2 repos themselves?

Can you point me to where are they in the `v1.x` tags? ;-)

We can backport these files to those branches but the logic would proven to be difficult as we check if the file exists, if it doesn't we fall back to the one in `test/k8sT/manifests/`

>Also, how will these be updated if the generated YAMLs are updated as well with new arguments, etc.?

These files shouldn't be changed because the arguments don't change. You might not have realized but the files in v1.2 are different from the files in v1.1 because there are arguments in v1.2 that don't exist in v1.1 for example.

ianvernon replied:
I'm not too set one way or another on this, so it's fine.

ianvernon commented on test/helpers/utils.go:
```diff
@@ -301,3 +304,48 @@ func DNSDeployment() string {
+	}
+	switch {
+	case CiliumV1_0.Check(cst):
+		return versioncheck.MustCompile(">= 1.8, <1.13"), nil
```
This should be 1.7 for v1.0? The CI tests for v1.0 test against v1.7 in `ginkgo.Jenkinsfile`. Same for v1.1.

aanm replied:
bad copy-paste, fixed

ianvernon reviewed (commented):
Awesome work, this is a great enhancement overall! I do have a few questions I would like answers to before giving full approval, though.
